File: python-examples/captcha-solving-auth-example/api/details.py
from typing import TypedDict

from intuned_browser import go_to_url
from playwright.async_api import BrowserContext, Page


class Params(TypedDict, total=False):
    name: str
    price: str
    details_url_item: str


class ProductDetails(TypedDict):
    name: str
    price: str
    sku: str
    category: str
    short_description: str
    full_description: str
    image_url: str
    sizes: list[str]
    colors: list[str]


async def automation(
    page: Page,
    params: Params | None = None,
    context: BrowserContext | None = None,
    **_kwargs,
):
    """
    Scrapes product details from a WooCommerce product page.

    Example params:
    {
        "name": "Chaz Kangeroo Hoodie",
        "price": "$52",
        "details_url_item": "https://scrapingcourse.com/ecommerce/product/chaz-kangeroo-hoodie"
    }
    """
    if not params or not params.get("details_url_item"):
        raise ValueError("details_url_item is required in params")

    url = params["details_url_item"]
    print(f"Navigating to: {url}")

    await go_to_url(page, url)

    # Basic info
    name = await page.locator('h1.product_title').text_content()
    price = await page.locator('.price .product-price').first.text_content()
    sku = await page.locator('.sku').text_content()
    category = await page.locator('.posted_in a').text_content()

    # Descriptions
    short_desc = await page.locator('.woocommerce-product-details__short-description p').text_content()
    full_desc_parts = []
    full_desc_locator = page.locator('#tab-description p')
    for i in range(await full_desc_locator.count()):
        text = await full_desc_locator.nth(i).text_content()
        if text:
            full_desc_parts.append(text.strip())
    full_description = '\n'.join(full_desc_parts)

    # Image
    image_url = await page.locator('.woocommerce-product-gallery__image a').first.get_attribute('href')

    # Sizes (skip "Choose an option")
    sizes = []
    size_options = page.locator('#size option')
    for i in range(await size_options.count()):
        value = await size_options.nth(i).get_attribute('value')
        if value:
            sizes.append(value)

    # Colors (skip "Choose an option")
    colors = []
    color_options = page.locator('#color option')
    for i in range(await color_options.count()):
        value = await color_options.nth(i).get_attribute('value')
        if value:
            colors.append(value)

    product: ProductDetails = {
        'name': name.strip() if name else '',
        'price': price.strip() if price else '',
        'sku': sku.strip() if sku else '',
        'category': category.strip() if category else '',
        'short_description': short_desc.strip() if short_desc else '',
        'full_description': full_description,
        'image_url': image_url or '',
        'sizes': sizes,
        'colors': colors,
    }

    print(f"Scraped product: {product['name']}")

    return product

File: typescript-examples/captcha-solving-auth/api/details.ts
import { BrowserContext, Page } from "playwright";
import { goToUrl } from "@intuned/browser";

interface Params {
  name: string;
  price: string;
  details_url_item: string;
}

interface ProductDetails {
  name: string;
  price: string;
  sku: string;
  category: string;
  short_description: string;
  full_description: string;
  image_url: string;
  sizes: string[];
  colors: string[];
}

export default async function handler(
  params: Params,
  page: Page,
  context: BrowserContext
): Promise<ProductDetails> {
  /**
   * Scrapes product details from a WooCommerce product page.
   *
   * Example params:
   * {
   *   "name": "Chaz Kangeroo Hoodie",
   *   "price": "$52",
   *   "details_url_item": "https://scrapingcourse.com/ecommerce/product/chaz-kangeroo-hoodie"
   * }
   */
  if (!params || !params.details_url_item) {
    throw new Error("details_url_item is required in params");
  }

  const url = params.details_url_item;
  console.log(`Navigating to: ${url}`);

  await goToUrl({ page, url });

  // Basic info
  const name = (await page.locator("h1.product_title").textContent()) || "";
  const price =
    (await page.locator(".price .product-price").first().textContent()) || "";
  const sku = (await page.locator(".sku").textContent()) || "";
  const category = (await page.locator(".posted_in a").textContent()) || "";

  // Descriptions
  const short_desc =
    (await page
      .locator(".woocommerce-product-details__short-description p")
      .textContent()) || "";

  const full_desc_parts: string[] = [];
  const full_desc_locator = page.locator("#tab-description p");
  const count = await full_desc_locator.count();
  for (let i = 0; i < count; i++) {
    const text = await full_desc_locator.nth(i).textContent();
    if (text) {
      full_desc_parts.push(text.trim());
    }
  }
  const full_description = full_desc_parts.join("\n");

  // Image
  const image_url =
    (await page
      .locator(".woocommerce-product-gallery__image a")
      .first()
      .getAttribute("href")) || "";

  // Sizes (skip "Choose an option")
  const sizes: string[] = [];
  const size_options = page.locator("#size option");
  const size_count = await size_options.count();
  for (let i = 0; i < size_count; i++) {
    const value = await size_options.nth(i).getAttribute("value");
    if (value) {
      sizes.push(value);
    }
  }

  // Colors (skip "Choose an option")
  const colors: string[] = [];
  const color_options = page.locator("#color option");
  const color_count = await color_options.count();
  for (let i = 0; i < color_count; i++) {
    const value = await color_options.nth(i).getAttribute("value");
    if (value) {
      colors.push(value);
    }
  }

  const product: ProductDetails = {
    name: name.trim(),
    price: price.trim(),
    sku: sku.trim(),
    category: category.trim(),
    short_description: short_desc.trim(),
    full_description,
    image_url,
    sizes,
    colors,
  };

  console.log(`Scraped product: ${product.name}`);

  return product;
}
